File: ios/UmbraCore/Sources/UmbraCore/Spec.swift
import Foundation

/// The platform-agnostic Umbra core, decoded from `umbra-core.json` (emitted by
/// the Python reference via `umbra export-spec`). iOS consumes exactly this — it
/// does not re-derive the posture model — so the phone and the laptop agree on
/// what a profile means and what this platform can honestly enforce.
///
/// Keep in step with umbra/spec.py::build_spec() and the Android core/Spec.kt.

/// The six honest promise levels. Mirrors umbra.platform.Enforcement.
public enum Enforcement: String, Codable {
    case enforced
    case requiresEntitlement = "requires_entitlement"
    case requiresVpnProfile = "requires_vpn_profile"
    case requiresRootedOS = "requires_rooted_os"
    case advisory
    case unavailable

    /// True when the app can actually change the state (vs. only observe/report).
    public var actionable: Bool {
        switch self {
        case .enforced, .requiresEntitlement, .requiresVpnProfile, .requiresRootedOS:
            return true
        case .advisory, .unavailable:
            return false
        }
    }

    public static func fromWire(_ s: String) -> Enforcement {
        Enforcement(rawValue: s) ?? .unavailable
    }
}

public struct EnforcementLevel: Codable {
    public let name: String
    public let description: String
}

public struct Capability: Codable {
    public let description: String
    public let controls: [String]
}

public struct PlatformCapability: Codable {
    public let level: String
    public let reason: String

    public var enforcement: Enforcement { Enforcement.fromWire(level) }
}

public struct PlatformSpec: Codable {
    public let description: String
    public let capabilities: [String: PlatformCapability]
}

public struct ProfileSpec: Codable {
    public let description: String
    public let failMode: String
    public let requires: [String]
    public let telemetryBlocklists: [String]

    enum CodingKeys: String, CodingKey {
        case description
        case failMode = "fail_mode"
        case requires
        case telemetryBlocklists = "telemetry_blocklists"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        description = try c.decodeIfPresent(String.self, forKey: .description) ?? ""
        failMode = try c.decodeIfPresent(String.self, forKey: .failMode) ?? "closed"
        requires = try c.decodeIfPresent([String].self, forKey: .requires) ?? []
        telemetryBlocklists = try c.decodeIfPresent([String].self, forKey: .telemetryBlocklists) ?? []
    }
}

public struct CoreSpec: Codable {
    public let specVersion: String
    public let engineVersion: String
    public let enforcementLevels: [EnforcementLevel]
    public let capabilities: [String: Capability]
    public let platforms: [String: PlatformSpec]
    public let telemetryBlocklists: [String: [String]]
    public let profiles: [String: ProfileSpec]

    enum CodingKeys: String, CodingKey {
        case specVersion = "umbra_spec_version"
        case engineVersion = "engine_version"
        case enforcementLevels = "enforcement_levels"
        case capabilities, platforms, profiles
        case telemetryBlocklists = "telemetry_blocklists"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        specVersion = try c.decode(String.self, forKey: .specVersion)
        engineVersion = try c.decode(String.self, forKey: .engineVersion)
        enforcementLevels = try c.decodeIfPresent([EnforcementLevel].self, forKey: .enforcementLevels) ?? []
        capabilities = try c.decodeIfPresent([String: Capability].self, forKey: .capabilities) ?? [:]
        platforms = try c.decodeIfPresent([String: PlatformSpec].self, forKey: .platforms) ?? [:]
        telemetryBlocklists = try c.decodeIfPresent([String: [String]].self, forKey: .telemetryBlocklists) ?? [:]
        profiles = try c.decodeIfPresent([String: ProfileSpec].self, forKey: .profiles) ?? [:]
    }

    /// The enforcement level THIS platform (ios) can promise for a capability.
    public func support(_ capability: String, platform: String = "ios") -> PlatformCapability {
        platforms[platform]?.capabilities[capability]
            ?? PlatformCapability(level: "unavailable", reason: "capability not in spec")
    }

    /// The de-duplicated union of telemetry domains a profile sinkholes.
    public func telemetryDomains(_ profileName: String) -> Set<String> {
        let lists = profiles[profileName]?.telemetryBlocklists ?? []
        return Set(lists.flatMap { telemetryBlocklists[$0] ?? [] })
    }

    public static func parse(_ data: Data) throws -> CoreSpec {
        try JSONDecoder().decode(CoreSpec.self, from: data)
    }

    /// Load the bundled copy shipped in the app/extension resources.
    public static func loadBundled(from bundle: Bundle = .main) throws -> CoreSpec {
        guard let url = bundle.url(forResource: "umbra-core", withExtension: "json") else {
            throw NSError(domain: "UmbraCore", code: 1,
                          userInfo: [NSLocalizedDescriptionKey: "umbra-core.json not in bundle"])
        }
        return try parse(Data(contentsOf: url))
    }
}

File: android/app/src/main/java/com/forrestlasiter/umbra/core/Spec.kt
package com.forrestlasiter.umbra.core

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * The platform-agnostic Umbra core, as shipped in `spec/umbra-core.json` (emitted
 * by the Python reference via `umbra export-spec`). Android does not re-derive the
 * posture model — it consumes exactly this, so the phone and the laptop agree on
 * what every profile means and what this platform can honestly enforce.
 *
 * Keep these types in step with umbra/spec.py::build_spec().
 */

/** The six honest promise levels. Mirrors umbra.platform.Enforcement. */
enum class Enforcement(val wire: String) {
    ENFORCED("enforced"),
    REQUIRES_ENTITLEMENT("requires_entitlement"),
    REQUIRES_VPN_PROFILE("requires_vpn_profile"),
    REQUIRES_ROOTED_OS("requires_rooted_os"),
    ADVISORY("advisory"),
    UNAVAILABLE("unavailable");

    /** True when the app can actually change the state (vs. only observe/report). */
    val actionable: Boolean
        get() = this == ENFORCED || this == REQUIRES_ENTITLEMENT ||
            this == REQUIRES_VPN_PROFILE || this == REQUIRES_ROOTED_OS

    companion object {
        fun fromWire(s: String): Enforcement =
            entries.firstOrNull { it.wire == s } ?: UNAVAILABLE
    }
}

@Serializable
data class EnforcementLevel(val name: String, val description: String)

@Serializable
data class Capability(
    val description: String,
    val controls: List<String> = emptyList(),
)

@Serializable
data class PlatformCapability(
    val level: String,
    val reason: String,
) {
    val enforcement: Enforcement get() = Enforcement.fromWire(level)
}

@Serializable
data class PlatformSpec(
    val description: String,
    val capabilities: Map<String, PlatformCapability> = emptyMap(),
)

@Serializable
data class ProfileSpec(
    val description: String = "",
    @SerialName("fail_mode") val failMode: String = "closed",
    val requires: List<String> = emptyList(),
    @SerialName("telemetry_blocklists") val telemetryBlocklists: List<String> = emptyList(),
)

@Serializable
data class CoreSpec(
    @SerialName("umbra_spec_version") val specVersion: String,
    @SerialName("engine_version") val engineVersion: String,
    @SerialName("enforcement_levels") val enforcementLevels: List<EnforcementLevel> = emptyList(),
    val capabilities: Map<String, Capability> = emptyMap(),
    val platforms: Map<String, PlatformSpec> = emptyMap(),
    @SerialName("telemetry_blocklists") val telemetryBlocklists: Map<String, List<String>> = emptyMap(),
    val profiles: Map<String, ProfileSpec> = emptyMap(),
) {
    /** The enforcement level THIS platform (android) can promise for a capability. */
    fun support(capability: String, platform: String = "android"): PlatformCapability =
        platforms[platform]?.capabilities?.get(capability)
            ?: PlatformCapability("unavailable", "capability not in spec")

    /** The de-duplicated union of telemetry domains a profile sinkholes. */
    fun telemetryDomains(profileName: String): Set<String> {
        val lists = profiles[profileName]?.telemetryBlocklists ?: emptyList()
        return lists.flatMap { telemetryBlocklists[it] ?: emptyList() }.toSet()
    }

    companion object {
        private val json = Json { ignoreUnknownKeys = true }

        fun parse(text: String): CoreSpec = json.decodeFromString(text)
    }
}
